### DISCUSSION TITLE
Get text between bookmarks

### DISCUSSION THREAD

**Discussion by SebastianSosa** [pymupdf/PyMuPDF#2177] [Looking for help]
Hi,

I am looking for a way to extract text between two bookmarks in a document. 
I have tried using a loop to go through each page and search for the exact match of a bookmark within each block of the page, but some bookmarks are within blocks of text (e.g. "Sample\nWe collected..." where Sample is a bookmark ). 
Currently, the only solution I can think of is to split all blocks based on "\n" and look for the exact match of a bookmark in each one.

But is there an easier way to extract text between bookmarks using PyMuPDF?

Thanks

**Comment by JorjMcKie** [comment]
If I am understanding correctly, you just need to extract the text with more detail. You seem to have been using `page.get_text("blocks")`, right?
If you use `page.get_text("dict")`, you have access to text "spans", which has the bboxes.
Then subselect spans with bbox top y-values greater than that of "Sample", ..

**Comment by SebastianSosa** [comment]
Indeed, I am using doc[0].get_textpage().extractBLOCKS() and, if I understand correctly, doc[0].get_textpage().extractDICT() returns each line of the page with bbox coordinates. 
Then, having the bbox of two bookmarks, can I extract all bbox that are between them?

Thank you

**Reply by JorjMcKie** [reply]
Correct - although there is a more user-friendly formulation of the same: `page.get_text("dict", ...)`.
You only want text, so use argument `flags=fitz.TEXTFLAGS_TEXT` instead of the 3 dots. Prevents any image extraction (will be faster, use less memory).
What is being returned is a stacked dictionary of dictionaries. The lowest level are so-called text "spans" - often a full line, multiple spans per line under special conditions like change in font properties, or if we have table cell content. Here is a snippet extracting lines between two vertical coordinates - in your case coming from two bookmarks.

So let's assume `top` is a float y-coordinate (derived from first bookmark) and `bot` the bottom float from bookmark 2.

You have two options - largely a matter of taste:

1. Extract all lines and check if their coordinates fall in between `top` and `bot`.
2. Extract lines only from within a "clip" rectangle, which has been built using `top`/`bot`.

Here is a snippet for option 1. The containment check ensures that only fully contained lines are considered. Obviously you can lower the bar in any way you want.
```python
# let "page" be the page where both bookmarks are pointing to, then
text = page.get_text("dict", flags=fitz.TEXTFLAGS_TEXT, sort=True)
for block in text["blocks"]:
    for line in block["lines"]:
        bbox = fitz.Rect(line["bbox"])  # the line rectangle
        if bbox.y1 < top or bbox.y0 > bot:  # line completely outside stripe defined by top/bot
            continue
        line_text = " ".join([span["text"].strip() for span in line["spans"]])
        print(line_text)
```

**Reply by SebastianSosa** [reply]
Perfect! Thank you very much!
